pkg/ctlog/config.go

```diff
@@ -0,0 +1,324 @@
+	}
+	if err := privKey.Validate(); err != nil {
+		return nil, fmt.Errorf("private key is not valid: %w", err)
+	}
+	ret.PrivKey = privKey
+
+	// If there's legacy rootCA entry, check it first.
+	if legacyRoot, ok := in[LegacyRootCAKey]; ok {
+		ret.FulcioCerts = append(ret.FulcioCerts, legacyRoot)
+	}
+	// Then loop through Fulcio roots
+	for k, v := range in {
+		if strings.HasPrefix(k, "fulcio-") {
+			ret.FulcioCerts = append(ret.FulcioCerts, v)
+		}
+	}
+	return &ret, nil
+}
+
+// MarshalConfig marshals the CTLogConfig into a format that can be handed
+// to the CTLog in form of a secret or configmap. Returns a map with the
+// following keys:
+// config - CTLog configuration
+// private - CTLog private key, PEM encoded and encrypted with the password
+// public - CTLog public key, PEM encoded
+// fulcio-%d - For each fulcioCerts, contains one entry so we can support
+// multiple.
+func (c *CTLogConfig) MarshalConfig(ctx context.Context) (map[string][]byte, error) {
+	// Since we can have multiple Fulcio secrets, we need to construct a set
+	// of files containing them for the RootsPemFile. Names don't matter
+	// so we just call them fulcio-%
+	// What matters however is to ensure that the filenames match the keys
+	// in the configmap / secret that we construc so they get properly mounted.
+	rootPems := make([]string, 0, len(c.FulcioCerts))
+	for i := range c.FulcioCerts {
+		rootPems = append(rootPems, fmt.Sprintf("%sfulcio-%d", rootsPemFileDir, i))
+	}
+	keyDER, err := x509.MarshalPKIXPublicKey(c.PrivKey.Public())
+	if err != nil {
+		logging.FromContext(ctx).Panicf("Failed to marshal the public key: %v", err)
+	}
+	proto := configpb.LogConfig{
+		LogId:        c.LogID,
+		Prefix:       c.LogPrefix,
+		RootsPemFile: rootPems,
+		PrivateKey: mustMarshalAny(&keyspb.PEMKeyFile{
+			Path:     privateKeyFile,
+			Password: c.PrivKeyPassword}),
+		PublicKey:      &keyspb.PublicKey{Der: keyDER},
+		LogBackendName: "trillian",
+		ExtKeyUsages:   []string{"CodeSigning"},
+	}
+
+	multiConfig := configpb.LogMultiConfig{
+		LogConfigs: &configpb.LogConfigSet{
+			Config: []*configpb.LogConfig{&proto},
+		},
+		Backends: &configpb.LogBackendSet{
+			Backend: []*configpb.LogBackend{{
+				Name:        "trillian",
+				BackendSpec: c.TrillianServerAddr,
+			}},
+		},
+	}
+	marshalledConfig, err := prototext.Marshal(&multiConfig)
+	secrets, err := c.marshalSecrets()
+	if err != nil {
+		return nil, err
+	}
+	secrets[ConfigKey] = marshalledConfig
+	return secrets, nil
+}
+
+// MarshalSecrets returns a map suitable for creating a secret out of
+// containing the following keys:
+// private - CTLog private key, PEM encoded and encrypted with the password
+// public - CTLog public key, PEM encoded
+// fulcio-%d - For each fulcioCerts, contains one entry so we can support
+// multiple.
+func (c *CTLogConfig) marshalSecrets() (map[string][]byte, error) {
+	// Encode private key to PKCS#1 ASN.1 PEM.
+	block := &pem.Block{
```

Comment:
change marshalling to "PRIVATE KEY" and MarshalPkcs8PrivateKey, and below change to "PUBLIC KEY" and MarshalPKIXPublicKey